File: src/content.rs
pub trait State {
    fn is_empty(&self) -> bool;
    fn is_contradiction(&self) -> bool;
}

pub trait Merge {
    fn merge(&self, other: &Self) -> Self;
}

#[derive(Hash, PartialEq, Eq, Clone)]
pub enum Content<A> {
    Nothing,
    Value(A),
    Contradiction
}

impl<A> Default for Content<A> {
    fn default() -> Self {
        Self::Nothing
    }
}

impl<A> State for Content<A> {
    fn is_empty(&self) -> bool {
        match self {
            Self::Nothing => true,
            _ => false
        }
    }

    fn is_contradiction(&self) -> bool {
        match self {
            Self::Contradiction => true,
            _ => false
        }
    }
}

impl<T> Content<T> {
    pub fn map_or<U, F: FnOnce(&T) -> U>(&self, default: U, f: F) -> U {
        match self {
            Self::Value(val) => f(&val),
            _ => default,
        }
    }
}

impl<T: Clone> Content<T> {
    pub fn map<F: FnOnce(&T) -> Self>(&self, f: F) -> Self {
        match self {
            Self::Value(val) => f(&val),
            _ => self.clone(),
        }
    }

    pub fn lift<F: Fn(&T, &T) -> Self>(&self, other: &Self, f: F) -> Self {
        match (self, other) {
            (Self::Nothing, Self::Nothing) => Self::Nothing,
            (Self::Value(val), Self::Nothing) => Self::Value(val.clone()),
            (Self::Nothing, Self::Value(val)) => Self::Value(val.clone()),
            (Self::Value(a), Self::Value(b)) => f(&a, &b),
            (_, Self::Contradiction) => Self::Contradiction,
            (Self::Contradiction, _) => Self::Contradiction
        }
    }
}

File: src/content.zig
const std = @import("std");
const warn = @import("std").debug.warn;

pub fn Generics(comptime T: type) type {
  return struct {
    merge: fn(*const T, *const T) ?T,
    eq: fn(T, T) bool = null,
    add: ?fn(T, T) ?T = null,
    sub: ?fn(T, T) ?T = null,
    mul: ?fn(T, T) ?T = null,
    less_than: ?fn(T, T) bool = null,
  };
}

pub fn Content(comptime T: type, fns: Generics(T)) type {
    return union(enum) {
        const Self = @This();

        Nothing,
        Value: T,
        Contradiction,

        pub fn init() Self {
            return .Nothing;
        }

        pub fn from(value: T) Self {
            return Self { .Value = value };
        }

        pub fn eq(self: Self, other: Self) bool {
            switch (self) {
                .Nothing => {
                    return other == .Nothing;
                },
                .Contradiction => {
                    return other == .Contradiction;
                },
                .Value => |self_value| {
                    switch (other) {
                        .Value => |other_value| {
                            return fns.eq(self_value, other_value);
                        },
                        else => { return false; }
                    }
                }
            }
        }

        pub fn less_than(self: Self, other: Self) bool {
            switch (self) {
                .Nothing => {
                    return false;
                },
                .Contradiction => {
                    //FIXME
                    if (other == .Value) { 
                        return true;
                    }

                    return false;
                },
                .Value => |self_value| {
                    switch (other) {
                        .Value => |other_value| {
                            if (fns.less_than) |f| {
                                return f(self_value, other_value);
                            }
                            else {
                                return self_value < other_value;
                            }
                        },
                        else => { return false; }
                    }
                }
            }
        }

        pub fn add(self: Self, other: Self) Self {
          if (fns.add) |f| {
            return lift(self, other, f);
          }
          else {
            return .Nothing;
          }
        }

        pub fn sub(self: Self, other: Self) Self {
          if (fns.sub) |f| {
            return lift(self, other, f);
          }
          else {
            return .Nothing;
          }
        }

        pub fn lift(self: Self, other: Self, f: fn(T, T) ?T) Self {
          switch (self) {
            .Value => |self_value| {
                switch (other) {
                    .Value => |other_value| {
                      const new_value = f(self_value, other_value);

                      if (new_value) |val| {
                        return Self { .Value = val };
                      }
                      else { 
                        return .Contradiction;
                      }
                    },
                    else => { 
                      return self;
                    }
                }
            },
            else => {
              return self;
            }
          }
        }

        pub fn is_empty(self: Self) bool {
          if (self == .Nothing) {
            return true;
          }

          return false;
        }

        pub fn clone(self: *const Self) Self {
            switch (self.*) {
                .Nothing => { 
                    return .Nothing;
                },
                .Value => |value| {
                    return Self { .Value = value };
                },
                .Contradiction => {
                    return .Contradiction;
                }
            }
        }

        pub fn merge(self: *const Self, other: *const Self) Self {
            switch (self.*) {
                .Nothing => { 
                    return other.clone(); 
                },
                .Value => |old_value| {
                  switch (other.*) {
                      .Nothing => { return .Nothing; },
                      .Value => |value| {
                          var new_value = fns.merge(&old_value, &value);

                          if (new_value) |val| {
                            return self.clone();
                          } else {
                            return Self.Contradiction;
                        }
                      },
                      .Contradiction => {
                          return .Contradiction;
                      }
                  }
                },
                .Contradiction => {
                    return .Contradiction;
                }
            }
        }
    };
}


